### PR TITLE
Fix hang if you invoke code fix during restore

### PR DESCRIPTION
**Pull request by davkean**
Fixes: https://github.com/dotnet/project-system/issues/5834.

We were not connecting the chain between NuGet restore operations that needed the UI thread, and Roslyn which was waiting on operation progress to complete. This is because our IProgressTrackerOutputDataSource didn't implement IJoinableProjectValueDataSource and connect it to the same collection that we're using to run/nominate restore.

This connects the dots between the two things, which required:

1) Moving IPackageRestoreService to derive from ProjectValueDataSource<T>
2) We no longer throw it way based on removal of PackageReference, and instead just prevent us from nominating NuGet. It is quite complex to have a data source that shuts itself down and stops producing data after a capability has been removed. We need a pattern for this, but this bug is not the one to do it in.

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+{
+    /// <summary>
+    ///     Responsible for pushing ("nominating") project data such as referenced packages and 
+    ///     target frameworks to NuGet so that it can perform a package restore and returns the 
+    ///     result.
+    /// </summary>
+    [Export(typeof(IPackageRestoreDataSource))]
+    [Export(ExportContractNames.Scopes.UnconfiguredProject, typeof(IProjectDynamicLoadComponent))]
+    [AppliesTo(ProjectCapability.PackageReferences)]
+    internal class PackageRestoreDataSource : ChainedProjectValueDataSourceBase<RestoreData>, IPackageRestoreDataSource, IProjectDynamicLoadComponent
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs]: Almost all of this code is identical with exception to LinkExternalInput/LoadAsync/UnloadAsync/RestoreAsync.

**Comment by davkean** [comment]
tag @lifengl

**Comment by davidwengier** [comment]
Should this target dev16.5.x?

**Comment by davkean** [comment]
Indeed.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs (hunk 1 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs]: Is there a pattern to extract here then?

Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs]: As in, I deleted PackageRestoreService and replaced it with this class but the diff doesn't show that.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        //                                             ||
+        //                                        RestoreData
+        //                                        //        \\ 
+        //                                       //          \\
+        //  ____________________________________\/___      ___\/____________________________________
+        // |                                         |    |                                         |
+        // |        PackageRestoreProgressTracker    |    |       PackageRestoreProgressTracker     |  Publishes restore progress to operation progress
+        // |            (Debug|AnyCPU|net45)         |    |       (Debug|AnyCPU|netcoreapp30)       |
+        // |_________________________________________|    |_________________________________________|
+        //
```
Comment (davidwengier) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs]: <3 <3 <3
🎉🎉 🎉  

Yay comments, and yay ascii art!

**Pull request merged**